qutebrowser/browser/webkit/network/pac.py

```diff
@@ -0,0 +1,299 @@
+            raise ParseProxyError(err.format(config[0]))
+
+    @staticmethod
+    def _parse_proxy_string(proxy_str):
+        proxies = proxy_str.split(';')
+        return [PACResolver._parse_proxy_entry(x) for x in proxies]
+
+    def __init__(self, pac_str):
+        """Create a PAC resolver.
+
+        Args:
+            pac_str: JavaScript code containing PAC resolver.
+        """
+        self._engine = QJSEngine()
+
+        self._ctx = _PACContext(self._engine)
+        self._engine.globalObject().setProperty("PAC",
+            self._engine.newQObject(self._ctx))
+        self._engine.evaluate(_PACContext.JS_DEFINITIONS, "pac_js_definitions")
+        self._engine.evaluate(utils.read_file("javascript/pac_utils.js"),
+            "pac_utils")
+        proxy_config = self._engine.newObject()
+        proxy_config.setProperty("bindings", self._engine.newObject())
+        self._engine.globalObject().setProperty("ProxyConfig", proxy_config)
+
+        self._engine.evaluate(pac_str, "pac")
+        global_js_object = self._engine.globalObject()
+        self._resolver = global_js_object.property("FindProxyForURL")
+        if not self._resolver.isCallable():
+            err = "Cannot resolve FindProxyForURL function, got '{}' instead"
+            raise EvalProxyError(err.format(self._resolver.toString()))
+
+    def resolve(self, query):
+        """Resolve a proxy via PAC.
+
+        Args:
+            query: QNetworkProxyQuery.
+
+        Return:
+            A list of QNetworkProxy objects in order of preference.
+        """
+        result = self._resolver.call([query.url().toString(),
+                                      query.peerHostName()])
+        result_str = result.toString()
+        if not result.isString():
+            err = "Got strange value from FindProxyForURL: '{}'"
+            raise EvalProxyError(err.format(result_str))
+        return self._parse_proxy_string(result_str)
+
+
+class PACFetcher(QObject):
+
+    """Asynchronous fetcher of PAC files."""
+
+    finished = pyqtSignal()
+
+    def __init__(self, url, parent=None):
+        """Resolve a PAC proxy from URL.
+
+        Args:
+            url: QUrl of a PAC proxy.
+        """
+        super().__init__(parent)
+
+        pac_prefix = "pac+"
+
+        assert url.scheme().startswith(pac_prefix)
+        url.setScheme(url.scheme()[len(pac_prefix):])
+
+        self._manager = QNetworkAccessManager()
+        self._manager.setProxy(QNetworkProxy(QNetworkProxy.NoProxy))
+        self._reply = self._manager.get(QNetworkRequest(url))
+        self._reply.finished.connect(self._finish)
+        self._pac = None
+        self._error_message = None
+
+    @pyqtSlot()
+    def _finish(self):
+        if self._reply.error() != QNetworkReply.NoError:
+            error = "Can't fetch PAC file from URL, error code {}: {}"
+            self._error_message = error.format(self._reply.error(),
+                self._reply.errorString())
```

Comment:
Same indenting thing as elsewhere :wink: Either align the second line with the `(`, or put both on the second line.